File: frontend/src/app/types/library.ts
import type { PageResponse } from "./knowledgeBase";

export type LibraryDocumentStatus = "active" | "disabled" | "archived";
export type LibraryParseJobStatus = "queued" | "running" | "success" | "failed" | "cancelled";
export type LibraryMemberPermissionLevel =
  | "read_only"
  | "document_manage"
  | "library_viewer"
  | "library_binder"
  | "library_editor"
  | "library_manager";

export interface LibraryDocumentDTO {
  documentId: string;
  ownerId: string;
  libraryId: string | null;
  name: string;
  sourceType: string;
  status: LibraryDocumentStatus;
  activeVersionId: string | null;
  activeVersionNo?: number | null;
  activeVersionFileName?: string | null;
  latestParseStatus?: string | null;
  latestParseRevisionId?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface LibraryDTO {
  libraryId: string;
  ownerId: string;
  ownerName: string;
  name: string;
  description: string | null;
  status: string;
  documentCount: number;
  createdAt: string;
  updatedAt: string;
}

export interface LibraryMemberDTO {
  bindingId: string;
  subjectType: "user" | "group";
  subjectId: string;
  permissionLevel: LibraryMemberPermissionLevel;
  status: string;
  createdAt: string;
}

export interface LibraryPageResponse {
  items: LibraryDTO[];
  total: number;
  pageNo: number;
  pageSize: number;
}

export interface ParseRevisionDTO {
  parseRevisionId: string;
  documentVersionId: string;
  status: "pending" | "running" | "success" | "failed" | "completed";
  contentFormat?: string;
  contentLength?: number;
  contentHash?: string | null;
  parserName: string | null;
  parserVersion?: string | null;
  parseOptions?: Record<string, unknown>;
  errorCode?: string | null;
  errorMessage?: string | null;
  createdAt: string;
  createdBy?: string | null;
}

export interface LibraryDocumentVersionDTO {
  versionId: string;
  documentId: string;
  versionNo: number;
  sourceFileId: string;
  fileName?: string;
  fileSize?: number;
  fileChecksum?: string | null;
  status: string;
  parseStatus: "pending" | "running" | "success" | "failed";
  chunkCount: number;
  tokenCount: number | null;
  createdAt: string;
  updatedAt: string;
  parseRevisions?: ParseRevisionDTO[];
}

export interface LibraryStoredFileDTO {
  fileId: string;
  fileName: string;
  mimeType: string | null;
  fileSize: number;
  checksum: string | null;
  objectKey: string;
}

export interface LibraryParseJobDTO {
  jobId: string;
  documentId: string;
  versionId: string;
  jobType: string;
  status: LibraryParseJobStatus;
  progress: number;
  errorCode: string | null;
  errorMessage: string | null;
  createdAt: string;
}

export interface LibraryDocumentUploadResponse {
  document: LibraryDocumentDTO;
  version: LibraryDocumentVersionDTO;
  parseJob: LibraryParseJobDTO;
  storedFile: LibraryStoredFileDTO;
}

export interface LibraryDocumentDetailDTO {
  document: LibraryDocumentDTO;
  activeVersion: LibraryDocumentVersionDTO | null;
}

export type LibraryDocumentPage = PageResponse<LibraryDocumentDTO>;

export interface LibraryTextPreviewResponse {
  text: string;
  truncated: boolean;
  fullLength: number;
}

export interface LibraryFullTextResponse {
  text: string;
}

export interface LibraryParsedChunkDTO {
  content: string;
  tokenCount: number;
  section?: string;
  pageNo?: number;
}

export interface LibraryParsedChunksResponse {
  chunks: LibraryParsedChunkDTO[];
}

export interface LibraryDocumentUsageDTO {
  bindingId: string;
  kbId: string;
  kbName: string;
  status: string;
  chunkCount: number;
  createdAt: string;
}

export interface LibraryDocumentUsageResponse {
  documentId: string;
  usages: LibraryDocumentUsageDTO[];
}

export interface LibraryStatsResponse {
  totalDocuments: number;
  todayUploads: number;
  pendingParse: number;
}

export interface BatchActionFailedItem {
  documentId: string;
  error: string;
  message: string;
}

export interface BatchActionSummary {
  total: number;
  succeeded: number;
  failed: number;
}

export interface BatchActionResponse {
  succeeded: string[];
  failed: BatchActionFailedItem[];
  summary: BatchActionSummary;
}

export interface UploadProgress {
  loaded: number;
  total: number;
  percent: number;
}

export interface UploadWithProgressResult {
  promise: Promise<LibraryDocumentUploadResponse>;
  cancel: () => void;
  onProgress: (callback: (progress: UploadProgress) => void) => void;
}

export interface LibraryVersionUploadResponse {
  version: LibraryDocumentVersionDTO;
  parseJob: LibraryParseJobDTO;
  storedFile: LibraryStoredFileDTO;
}

export interface LibraryVersionActivateResponse {
  documentId: string;
  activeVersionId: string;
  previousActiveVersionId: string | null;
}

export interface LibraryReparseRequest {
  parserName?: string | null;
  parserVersion?: string | null;
  contentFormat: "markdown" | "text";
  parseOptions: Record<string, unknown>;
  reason?: string | null;
}

export interface LibraryParseRevisionCreateResponse {
  jobId: string;
  parseRevisionId: string;
  status: string;
}

export interface DeletionImpactAnalysis {
  canDelete: boolean;
  blockingReasons: string[];
  isActiveVersion: boolean;
  activeBindingCount: number;
  pendingJobsCount: number;
  qaEvidenceCount: number;
  qaCitationCount: number;
  requiresStrongConfirmation: boolean;
}

export interface LibraryBindingDTO {
  bindingId: string;
  documentId: string;
  documentName: string;
  kbId: string;
  versionId: string;
  chunkSize: number;
  chunkOverlap: number;
  status: string;
  chunkCount: number;
  errorCode?: string | null;
  errorMessage?: string | null;
  activeChunkRevisionId?: string | null;
  chunkRevisionStatus?: string | null;
  chunkRevisionChunkCount?: number | null;
  chunkRevisionVersionId?: string | null;
  createdAt: string;
  createdBy?: string | null;
}

File: backend/app/schemas/library.py
from typing import Literal
from pydantic import BaseModel, Field


class LibraryDocumentDTO(BaseModel):
    """文档库文档主对象 DTO。"""

    documentId: str
    ownerId: str
    libraryId: str | None = None
    libraryName: str | None = None
    name: str
    sourceType: str
    status: str
    activeVersionId: str | None
    activeVersionNo: int | None = None
    activeVersionFileName: str | None = None
    latestParseStatus: str | None = None
    latestParseRevisionId: str | None = None
    createdAt: str
    updatedAt: str


class LibraryDocumentVersionDTO(BaseModel):
    """文档库源文件版本 DTO。"""

    versionId: str
    documentId: str
    versionNo: int
    sourceFileId: str
    fileName: str | None = None
    fileSize: int | None = None
    fileChecksum: str | None = None
    status: str
    parseStatus: str
    tokenCount: int | None
    activeParseRevisionId: str | None = None
    createdAt: str
    updatedAt: str


class LibraryStoredFileDTO(BaseModel):
    """文档库文件元数据 DTO。"""

    fileId: str
    fileName: str
    mimeType: str | None
    fileSize: int
    checksum: str | None
    objectKey: str


class LibraryParseJobDTO(BaseModel):
    """文档库解析作业 DTO。"""

    jobId: str
    documentId: str
    versionId: str
    jobType: str
    status: str
    progress: int
    errorCode: str | None
    errorMessage: str | None
    createdAt: str


class LibraryParseRevisionDTO(BaseModel):
    """文档库解析版本 DTO，不包含知识库分块数量。"""

    parseRevisionId: str
    documentVersionId: str
    status: str
    contentFormat: str
    contentLength: int
    contentHash: str | None
    parserName: str | None
    parserVersion: str | None
    parseOptions: dict
    errorCode: str | None = None
    errorMessage: str | None = None
    isActive: bool = False
    createdAt: str
    createdBy: str | None = None


class LibraryReparseRequest(BaseModel):
    """创建解析版本请求。"""

    parserName: str | None = "auto"
    parserVersion: str | None = None
    contentFormat: Literal["markdown", "text"] = "markdown"
    parseOptions: dict = Field(default_factory=dict)
    reason: str | None = None


class LibraryUploadParseOptions(BaseModel):
    """上传文档/版本时的解析选项。"""

    parserName: str | None = "auto"
    contentFormat: Literal["markdown", "text"] = "markdown"
    parseOptions: dict | None = None


class LibraryParseRevisionCreateResponse(BaseModel):
    """创建解析版本后的排队响应。"""

    jobId: str
    parseRevisionId: str
    status: str


class LibraryDocumentUploadResponse(BaseModel):
    """文档库上传成功响应。"""

    document: LibraryDocumentDTO
    version: LibraryDocumentVersionDTO
    parseJob: LibraryParseJobDTO
    storedFile: LibraryStoredFileDTO


class LibraryDocumentDetailDTO(BaseModel):
    """文档库文档详情响应。"""

    document: LibraryDocumentDTO
    activeVersion: LibraryDocumentVersionDTO | None


class LibraryDocumentUpdateRequest(BaseModel):
    """文档库文档更新请求。"""

    name: str | None = None
    status: str | None = None


class LibraryTextPreviewResponse(BaseModel):
    """文本预览响应。"""
    text: str
    truncated: bool
    fullLength: int


class LibraryParsedChunkDTO(BaseModel):
    """解析后的分块数据。"""
    content: str
    tokenCount: int
    section: str | None = None
    pageNo: int | None = None
    startOffset: int | None = None
    endOffset: int | None = None


class LibraryFullTextResponse(BaseModel):
    """完整文本响应。"""
    text: str


class LibraryParsedChunksResponse(BaseModel):
    """结构化解析分块响应。"""
    chunks: list[LibraryParsedChunkDTO]


class LibraryDocumentUsageDTO(BaseModel):
    """文档使用情况：绑定的知识库列表。"""
    bindingId: str
    kbId: str
    kbName: str
    status: str
    chunkCount: int
    createdAt: str


class LibraryDocumentUsageResponse(BaseModel):
    """文档使用情况响应。"""
    documentId: str
    usages: list[LibraryDocumentUsageDTO]


class BatchActionRequest(BaseModel):
    """批量操作请求。"""
    documentIds: list[str] = Field(..., min_length=1, max_length=100)
    action: Literal["delete", "reparse", "disable"]


class BatchActionFailedItem(BaseModel):
    """批量操作失败项。"""
    documentId: str
    error: str
    message: str


class BatchActionSummary(BaseModel):
    """批量操作汇总。"""
    total: int
    succeeded: int
    failed: int


class BatchActionResponse(BaseModel):
    """批量操作响应。"""
    succeeded: list[str]
    failed: list[BatchActionFailedItem]
    summary: BatchActionSummary


class LibraryStatsResponse(BaseModel):
    """文档库统计响应。"""
    totalDocuments: int
    todayUploads: int
    pendingParse: int


class LibraryVersionUploadResponse(BaseModel):
    """上传新版本成功响应。"""
    version: LibraryDocumentVersionDTO
    parseJob: LibraryParseJobDTO
    storedFile: LibraryStoredFileDTO


class LibraryVersionActivateRequest(BaseModel):
    """切换活跃版本请求。"""
    confirmImpact: bool = False


class LibraryVersionActivateResponse(BaseModel):
    """切换活跃版本响应。"""
    documentId: str
    activeVersionId: str
    previousActiveVersionId: str | None


class LibraryParseRevisionActivateRequest(BaseModel):
    """切换活动解析修订请求。"""
    parseRevisionId: str


class LibraryParseRevisionActivateResponse(BaseModel):
    """切换活动解析修订响应。"""
    documentId: str
    versionId: str
    activeParseRevisionId: str | None
    previousActiveParseRevisionId: str | None
